### PR TITLE
Load GC Articles provided content but keep named routes

### PR DESCRIPTION
**Pull request by dsamojlenko**
# Summary | Résumé

Swap out static pages for GC Articles provided content, preserving named routes. 

The simpler version of this would be to delete the routes and just let the path pattern fall through to the wildcard route, but in some cases we want to preserve the named routes for use in templates. 

To preserve these named routes, we'll just leave them in place but use the page_content() method of the wildcard route to render the page.

### CODE REVIEW

**Comment by amazingphilippe** [comment]
Just fixed a few things. Here's the list:

1. Removed conflicting named routes
2. replaced `url_for()` with `_()` and added translations where routes were referenced. 
3. Fixed the personalisation-guide fr redirect

Just realized we did not need to update all `url_for`s. When the routes are redirecting, we can use the named routes just fine.

File: app/main/views/index.py [hunk 1]
```diff
@@ -95,11 +89,6 @@ def verify_mobile():
     return render_template("views/verify-mobile.html")
 
 
-@main.route("/privacy")
```
Comment (andrewleith) [inline app/main/views/index.py:98]: Why did you remove some routes entirely to allow pass-through to GCArticles (like this one), but then modify others to explicitly check the language (like `/format`)?  What is the difference with these two approaches?

File: app/templates/partials/nav/footer.html [hunk 2]
```diff
@@ -14,8 +14,8 @@
               {% endif %}
             {% endfor %}
           {% else %}
-            {{ nav_link_dark(id_key='nav-footer-primary-why-notify', label=_('Why GC Notify'), href=url_for('main.why-notify'), class="-ml-2") }}
-            {{ nav_link_dark(id_key='nav-footer-primary-features', label=_('Features'), href=url_for('main.features'), class="-ml-2") }}
+            {{ nav_link_dark(id_key='nav-footer-primary-why-notify', label=_('Why GC Notify'), href=_('/why-notify'), class="-ml-2") }}
```
Comment (andrewleith) [inline app/templates/partials/nav/footer.html:17]: This seems to work well for app-based pages.  However, when I am viewing static content on GCArticles, the french menus all have English URLs.  I tried to update them but it didn't seem to work.

**Review by andrewleith — COMMENTED** [review]
Overall this looks good.  I am curious what @jimleroyer thinks about having 2 ways to link to URLs going forward:
1. For app pages, we would continue to use`url_for()` as usual
2. For static content served up from GCArticles, we would create translations, and link directly using `<a href="_('/some-path-in-translation-file')">...</a>`

File: app/main/views/index.py (hunk 1 above)
Comment (amazingphilippe) [reply app/main/views/index.py:98]: That's a good point.
- I left the `/format` route and alike because we were forwarding it to a new path in both languages: `/formatting-guide` and `guides-reference`. My rationale is that we might want to keep these ones to redirect users if we forget a link somewhere. Basically preventing link rot. 
- I removed the `privacy` routes because they were causing problem when linking from FR pages. Instead of adding a new route, or doing something hacky in the route, we decided to remove it entirely and use relative path strings that we would manually translate.

File: app/templates/partials/nav/footer.html (hunk 2 above)
Comment (amazingphilippe) [reply app/templates/partials/nav/footer.html:17]: You're right, I can see the French menu in Articles, but it doesn't seem to get picked up on fr pages. This might already be an existing issue

File: app/main/views/index.py (hunk 1 above)
Comment (andrewleith) [reply app/main/views/index.py:98]: Gotcha - thanks for the explanation.

File: .github/workflows/test-admin-deploy.yaml [hunk 3]
```diff
@@ -132,7 +132,8 @@ jobs:
                 REDIS_ENABLED=False,\
                 SENDING_DOMAIN=staging.notification.cdssandbox.xyz,\
                 API_HOST_NAME=https://api.staging.notification.cdssandbox.xyz,\
-                ADMIN_BASE_URL=$URL
+                ADMIN_BASE_URL=$URL,\
+                GC_ARTICLES_API=articles.alpha.canada.ca/notification-gc-notify
```
Comment (jimleroyer) [inline .github/workflows/test-admin-deploy.yaml:136]: Spoke with Andrew, we'll undo this and go through the manifest .env files.

File: app/main/views/index.py [hunk 4]
```diff
@@ -48,14 +49,8 @@ def index():
     if current_user and current_user.is_authenticated:
         return redirect(url_for("main.choose_account"))
 
-    return render_template(
-        "views/signedout.html",
```
Comment (jimleroyer) [inline app/main/views/index.py:52]: One thing I just noticed is that we changed the hierarchy of the template. Initially, the notify_template.html was the base template for public pages and the main_template.html was the one for authenticated pages. It seems we switched to the latter, following the call hierarchy and template called by `page_content`. 

I wonder what we lose an gain from switching to the main_template.html file. I know that @andrewleith had to add back some metatags and the hidden ascii art easter egg already.

image

File: app/main/views/index.py [hunk 5]
```diff
@@ -71,7 +66,7 @@ def robots():
 
 @main.route("/.well-known/security.txt")
 def security_txt():
-    security_policy = url_for("main.security", _external=True)
+    security_policy = url_for("main.index", _external=True).rstrip("/") + _("/security")
```
Comment (jimleroyer) [inline app/main/views/index.py:69]: Maybe one slight improvement would be to have a util function in the `articles` module, something called `gc_url_for` for example that we could call this way:

```python
gc_url_for('/security')
```

File: app/main/views/index.py [hunk 6]
```diff
@@ -374,6 +318,16 @@ def old_page_redirects():
     return redirect(url_for(request.endpoint.replace("redirect_", "")), code=301)
 
 
+# --- GCA Redirects --- #
+@main.route("/a11y", endpoint="/accessibility")
+@main.route("/why-notify", endpoint="/why-gc-notify")
+@main.route("/personalise", endpoint="/personalisation-guide")
+@main.route("/format", endpoint="/formatting-guide")
+@main.route("/messages-status", endpoint="/formatting-guide")
```
Comment (jimleroyer) [inline app/main/views/index.py:326]: Note: message-status is redirecting to formatting-guide, check if that's right.

File: app/main/views/index.py [hunk 7]
```diff
@@ -374,6 +318,16 @@ def old_page_redirects():
...
+def gca_redirects():
+    return redirect(_(request.endpoint.replace("main.", "")), code=301)
```
Comment (jimleroyer) [inline app/main/views/index.py:328]: Nice handy function!

File: app/templates/content_template.html [hunk 8]
```diff
@@ -6,6 +6,22 @@
...
+      <meta name="keywords" content="{{ keywords }}"/>
+      <meta property="og:title" content="{{ meta_title }}"/>
+      <meta property="og:description" content="{{ description }}"/>
+      <meta
+        property="og:image"
+        content="{{ asset_url('images/product/notify-main.jpg') }}"
+      />
+      <meta property="og:url" content="{{ admin_base_url }}"/>
+    {% endif %}
+  {% endblock %}
```
Comment (jimleroyer) [inline app/templates/content_template.html:24]: If we'd be using the notify_template.html file as the base hierarchy instead of main_template.html, filling in the gap with the meta might not be necessary.

File: app/translations/csv/fr.csv [hunk 9]
```diff
@@ -1561,4 +1561,14 @@
...
+"/privacy","/confidentialite"
+"/security","/securite"
+"/why-gc-notify","/pourquoi-gc-notification"
+"/features","/fonctionnalites"
+"/guidance","/guides-reference"
+"/accessibility","/accessibilite"
+"/terms","/conditions-dutilisation"
+"/formatting-guide","/guide-mise-en-forme"
+"/personalisation-guide","/guide-personnalisation"
+"/message-delivery-status","/etat-livraison-messages"
```
Comment (jimleroyer) [inline app/translations/csv/fr.csv:1574]: Moving to a hypothetical `gca_url_for` function might remove the need to be in the translation file, but it would require a dictionary making this mapping. It would move content related translations from bilingual URLs. I wonder if placing such mapping in the navigation.py might be better. /brainstorming

File: tests/app/articles/test_pages.py [hunk 10]
```diff
@@ -99,3 +100,49 @@ def test_bad_slug_doesnt_save_empty_cache_entry(app_, mocker):
...
+    """
+    This test ensures that the pages we renamed properly provide a permanent redirect, and exist in GCA
+    """
+
+    # ensure each url is a permenent redirect
+    client_request.get_url(url, _expected_status=301)
+
+    # ensure target location exists and returns content
+    page = client_request.get_url(url, _follow_redirects=True)
+    assert page.find("span", id="gc-title").text.strip() == "GC Notify"
```
Comment (jimleroyer) [inline tests/app/articles/test_pages.py:116]: Very nice. 👍

File: tests/app/main/views/test_index.py [hunk 11]
```diff
@@ -96,7 +96,8 @@ def test_security_txt(client):
     response = client.get(url_for("main.security_txt"))
     assert response.headers["Content-Type"] == "text/plain"
     assert response.status_code == 200
-    security_policy = url_for("main.security", _external=True)
+    security_policy = url_for("main.index", _external=True).rstrip("/") + "/security"
```
Comment (jimleroyer) [inline tests/app/main/views/test_index.py:99]: We could use the hypothetical `gca_url_for`.

File: tests/app/main/views/test_index.py [hunk 12]
```diff
@@ -136,6 +128,29 @@ def test_static_pages(
...
+        "/message-delivery-status",
+    ],
+)
+def test_static_pages_gca(
+    client_request,
+    mock_get_organisation_by_domain,
+    url,
+):
+    page = client_request.get_url(url, _follow_redirects=True)
+    assert not page.select_one("meta[name=description]")
```
Comment (jimleroyer) [inline tests/app/main/views/test_index.py:151]: ♥

File: app/__init__.py [hunk 13]
```diff
@@ -508,7 +507,9 @@ def format_notification_status_as_field_status(status, notification_type):
 
 
 def format_notification_status_as_url(status, notification_type):
-    url = partial(url_for, "main.messages_status")
+    # url = partial(url_for, "main.messages_status")
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply app/__init__.py:510]: Done.

File: tests/app/main/views/test_index.py (hunk 11 above)
Comment (andrewleith) [reply tests/app/main/views/test_index.py:99]: Done.

File: .github/workflows/test-admin-deploy.yaml (hunk 3 above)
Comment (andrewleith) [reply .github/workflows/test-admin-deploy.yaml:136]: Done.

File: app/main/views/index.py (hunk 5 above)
Comment (andrewleith) [reply app/main/views/index.py:69]: Done.

File: app/templates/content_template.html (hunk 8 above)
Comment (andrewleith) [reply app/templates/content_template.html:24]: Will look into this next.

File: app/main/views/index.py (hunk 6 above)
Comment (andrewleith) [reply app/main/views/index.py]: Fixed.

File: app/templates/partials/nav/footer.html (hunk 2 above)
Comment (andrewleith) [reply app/templates/partials/nav/footer.html]: I think we are all set on this now, menus are now working as expected.

File: app/main/views/index.py (hunk 4 above)
Comment (andrewleith) [reply app/main/views/index.py:52]: @jimleroyer  Digging a little further, it looks like only `signedout.html` used `notify_template.html` - every other page used `admin_template.html` even pages that are not signed in.  We are not introducing anything new here in regards to templates inheritance.

**Comment by andrewleith** [comment]
Looking into getting the ASCII art working again on the home page, I noticed the following condition in `app/templates/partials/cds_ascii_art.html`:

```
{% if is_cds_domain or is_review_app %}
```

Not sure why we need that, but it just means you won't see it when running locally.  I've left it there for now.

File: tests/app/articles/test_pages.py [hunk 14]
```diff
@@ -99,3 +100,49 @@ def test_bad_slug_doesnt_save_empty_cache_entry(app_, mocker):
...
+    """
+    This test is ensures the content we link to exists on GCA.  If someone removes a page on GCA, this will fail.
+    """
+
+    # ensure target location exists and returns content
+    page = client_request.get_url(url, _expected_status=200)
+    assert (
+        page.find("span", id="gc-title").text.strip() == "GC Notify"
+        or page.find("span", id="gc-title").text.strip() == "GC Notification"
+    )
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply tests/app/articles/test_pages.py:148]: Yes, probably.  Shall I remove these tests?

File: app/articles/routing.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
+from flask import current_app, url_for
+
+from app.articles import get_current_locale
+
+GC_ARTICLES_ROUTES = {
+    "home": {"en": "/home", "fr": "/acceuil"},
```
Comment (jimleroyer) [inline app/articles/routing.py:6]: Is there a typo in `accueil` or it's really written this way in the app?

Comment (jimleroyer) [reply app/articles/routing.py:6]:
```suggestion
    "home": {"en": "/home", "fr": "/accueil"},
```

Comment (andrewleith) [reply app/articles/routing.py:6]: That is a typo.  I will fix and add a test to verify the URLs in this dict 😊

File: tests/app/main/views/test_index.py [hunk 16]
```diff
@@ -96,7 +97,7 @@ def test_security_txt(client):
     response = client.get(url_for("main.security_txt"))
     assert response.headers["Content-Type"] == "text/plain"
     assert response.status_code == 200
-    security_policy = url_for("main.index", _external=True).rstrip("/") + "/security"
```
Comment (jimleroyer) [inline tests/app/main/views/test_index.py]: 👍

File: .github/workflows/test-admin-deploy.yaml [hunk 17]
```diff
@@ -132,8 +132,7 @@ jobs:
                 REDIS_ENABLED=False,\
                 SENDING_DOMAIN=staging.notification.cdssandbox.xyz,\
                 API_HOST_NAME=https://api.staging.notification.cdssandbox.xyz,\
-                ADMIN_BASE_URL=$URL,\
-                GC_ARTICLES_API=articles.alpha.canada.ca/notification-gc-notify
```
Comment (jimleroyer) [inline .github/workflows/test-admin-deploy.yaml]: 💥

**Review by jimleroyer — COMMENTED** [review]
There might be a typo about accueil but else it looks excellent.

File: tests/app/articles/test_pages.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jimleroyer) [reply tests/app/articles/test_pages.py:148]: We could add an annotation for future usage and not lose track of these, so we could do something along the lines of what we did for [concurrent vs serial tests](https://github.com/cds-snc/notification-api/blob/ff0b04a435bde73bdf4815c06292a9da439fca82/scripts/run_tests.sh#L43-L48). But we'd just mark these for now without taking more actions on how to run. Pytest documentation: https://docs.pytest.org/en/6.2.x/mark.html

File: app/main/views/index.py (hunk 4 above)
Comment (jimleroyer) [reply app/main/views/index.py:52]: Thanks @andrewleith !

File: tests/app/articles/test_pages.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (andrewleith) [reply tests/app/articles/test_pages.py:148]: Are you suggesting marking them as `skip` - or marking them as integration tests and letting them continue to run?

File: app/articles/routing.py (hunk 15 above)
Comment (andrewleith) [reply app/articles/routing.py]: Fixed.

**Review by jimleroyer — APPROVED** [review]
LGTM! :shipit:
